scipy/linalg/_solvers.py

```diff
@@ -329,44 +343,81 @@ def solve_discrete_are(a, b, q, r):
-                         'is ill-conditioned')
-
-    g = np.dot(np.dot(b, g), b.conj().transpose())
-
-    try:
-        ait = inv(a).conj().transpose()  # ait is "A inverse transpose"
-    except LinAlgError:
-        raise ValueError('Matrix A in the algebraic Riccati equation solver '
-                         'is ill-conditioned')
-
-    z11 = a+np.dot(np.dot(g, ait), q)
-    z12 = -1.0*np.dot(g, ait)
-    z21 = -1.0*np.dot(ait, q)
-    z22 = ait
-
-    z = np.vstack((np.hstack((z11, z12)), np.hstack((z21, z22))))
-
-    # Note: we need to sort the upper left of s to lie within the unit circle,
-    #       while the lower right is outside (Laub, p. 7)
-    s, u, _ = schur(z, sort='iuc')
-
-    (m, n) = u.shape
-
-    u11 = u[0:m//2, 0:n//2]
-    u21 = u[m//2:m, 0:n//2]
-    u11i = inv(u11)
-
-    return np.dot(u21, u11i)
+    
+    a = _asarray_validated(a)
+    b = _asarray_validated(b)
+    q = _asarray_validated(q)
+    r = _asarray_validated(r)
+    
+    # Get the correct data types otherwise Numpy complains about pushing 
+    # complex numbers into real arrays.
+    r_or_c = complex if np.iscomplexobj(b) else float
+
+    for ind, x in enumerate((a,q,r)):
+        if np.iscomplexobj(x):
+            r_or_c = complex
+        if not np.equal(*x.shape):
+            raise ValueError("Matrix {} should be square.".format("aqr"[ind]))
+    
+    # Shape consistency checks
+    m, n = b.shape
+
+    if m != a.shape[0]:
+        raise ValueError("Matrix a and b should have the same number of rows.")
+        
+    if m != q.shape[0]:
+        raise ValueError("Matrix a and q should have the same shape.")
+            
+    if n != r.shape[0]:
+        raise ValueError("Matrix b and r should have the same number of cols.")
+    
+    # Check if the data is (sufficiently) hermitian
+    if norm(q - q.conj().T,1) > np.spacing(norm(q,1))*100:
+        raise ValueError("Matrix q should be symmetric/hermitian.")
+
+    if norm(r - r.conj().T,1) > np.spacing(norm(r,1))*100:
+        raise ValueError("Matrix r should be symmetric/hermitian.")
+
+    # Form the matrix pencil        
+    H = np.zeros((2*m + n, 2*m + n), dtype = r_or_c)
+    H[:m, :m] = a
+    H[:m, 2*m:] = b
+    H[m:2*m, :m] = -q
+    H[m:2*m, m:2*m] = np.eye(m)
+    H[2*m:, 2*m:] = r
+
+    J = np.zeros_like(H, dtype = r_or_c)
+    J[:m, :m] = np.eye(m)
+    J[m:2*m, m:2*m] = a.conj().T
+    J[2*m:, m:2*m] = -b.conj().T
+
+    # Deflate the pencil by the R column
+    q_of_qr, _ = qr(H[:,-n:])
+    H = q_of_qr[:,n:].conj().T.dot(H[:,:2*m])
+    J = q_of_qr[:,n:].conj().T.dot(J[:,:2*m])
+    
+    # Decide on which output type is needed for QZ 
+    out_str = 'real' if r_or_c == float else 'complex'
+    
+    _, _, _, _, _, u = ordqz(H, J, sort = 'iuc', overwrite_a=True, 
+                    overwrite_b=True, check_finite=False, output=out_str)
+    
+    # Get the relevant parts of the stable subspace basis
+    # TODO: Check if it succeded to get the right ordering.
```

Comment:
These equations don't have solvability guarantees. Depending on the data matrices a solution might not exist. I don't know what SciPy way when no solution is found : return empty matrix or raise exception ?